kratos/processes/compute_nodal_gradient_process.cpp

```diff
@@ -32,77 +33,132 @@ void ComputeNodalGradientProcess<THistorical>::Execute()
     ClearGradient();
 
     // Auxiliar containers
-    Matrix DN_DX, J0;
-    Vector N;
+    Matrix DN_DX, J0, InvJ0;
 
     // First element iterator
     const auto it_element_begin = mrModelPart.ElementsBegin();
 
     // Current domain size
     const std::size_t dimension = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
 
-    // Iterate over the elements
-    #pragma omp parallel for firstprivate(DN_DX,  N, J0)
-    for(int i_elem=0; i_elem<static_cast<int>(mrModelPart.Elements().size()); ++i_elem) {
-        auto it_elem = it_element_begin + i_elem;
-        auto& r_geometry = it_elem->GetGeometry();
-
-        // Current geometry information
-        const std::size_t local_space_dimension = r_geometry.LocalSpaceDimension();
-        const std::size_t number_of_nodes = r_geometry.PointsNumber();
-
-        // Resize if needed
-        if (DN_DX.size1() != number_of_nodes || DN_DX.size2() != dimension)
-            DN_DX.resize(number_of_nodes, dimension);
-        if (N.size() != number_of_nodes)
-            N.resize(number_of_nodes);
-        if (J0.size1() != dimension || J0.size2() != local_space_dimension)
-            J0.resize(dimension, local_space_dimension);
-
-        // The integration points
-        const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod();
-        const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method);
-        const std::size_t number_of_integration_points = r_integration_points.size();
-
-        Vector values(number_of_nodes);
-        if (!mNonHistoricalVariable) {
```

Comment:
In order to avoid this check you are hard copying the whole code, what is super error prone and not recommended. Please think a better solutions that does not require hardcopy

Reply:
well, but the if should definitely not be inside the loop

Reply:
Since it is a technical issue, I have nothing to comment on.
So, maybe you can discuss it later as it suits you better.

Reply:
Fixed with an auxiliar struct